### PR TITLE
Test-DbaMaxMemory - Added warning and filter for engine

### LINKED ISSUE #8249: Bug in Test-DbaMaxMemory with InstanceCount?
### Verified issue does not already exist?

Yes

### What error did you receive?

No error...see "Steps to reproduce" section.

### Steps to Reproduce

A few times now I have run the the Test-DbaMaxMemory cmdlet on some systems that have a single instance on them, but this cmdlet tells me I have 2 instances and recommends memory based on that. I dug into the code here in the repo, and I believe there is a bug on this line:

$instanceCount = ($serverService | Where-Object State -Like Running | Where-Object InstanceName | Group-Object InstanceName | Measure-Object Count).Count

I'll explain. The code first calls:

$serverService = Get-DbaService -ComputerName $instance -EnableException

And on the system I ran this on today, it returned the following:

ComputerName : TestServer
ServiceName : SQLServerReportingServices
ServiceType : SSRS
InstanceName : SSRS
DisplayName : SQL Server Reporting Services
StartName : NT SERVICE\SQLServerReportingServices
State : Running
StartMode : Automatic

ComputerName : TestServer
ServiceName : MsDtsServer150
ServiceType : SSIS
InstanceName :
DisplayName : SQL Server Integration Services 15.0
StartName : company\account
State : Running
StartMode : Automatic

ComputerName : TestServer
ServiceName : MSSQLSERVER
ServiceType : Engine
InstanceName : MSSQLSERVER
DisplayName : SQL Server (MSSQLSERVER)
StartName : company\account
State : Running
StartMode : Automatic

ComputerName : TestServer
ServiceName : SQLBrowser
ServiceType : Browser
InstanceName :
DisplayName : SQL Server Browser
StartName : NT AUTHORITY\LOCALSERVICE
State : Stopped
StartMode : Disabled

ComputerName : TestServer
ServiceName : SQLSERVERAGENT
ServiceType : Agent
InstanceName : MSSQLSERVER
DisplayName : SQL Server Agent (MSSQLSERVER)
StartName : company\account
State : Running
StartMode : Automatic

This line is then called:

$instanceCount = ($serverService | Where-Object State -Like Running | Where-Object InstanceName | Group-Object InstanceName | Measure-Object Count).Count

It groups on the InstanceName, and since there are two MSSQLSERVER objects that are running (the engine and SQL agent), it thinks there are two instances (if I am reading this right).

If we really wanted to know how many instances there are, wouldn't we want to know how many instances of the ENGINE there is, and therefore filter on ServiceType = Engine?

$instanceCount = ($serverService | Where-Object State -Like Running | Where-Object InstanceName | Where-Object ServiceType -Like Engine | Group-Object InstanceName | Measure-Object Count).Count

Ultimately, I have run this a few systems where I know there is a single SQL Server instance but this cmdlet returns a value of 2 for the instance count.

Interestingly enough, for further testing, I stopped the SQL Server agent and reran things and it still returned a value of 2. So maybe we could get away with something like this:

$instanceCount = ($serverService | Where-Object State -Like Running | Where-Object ServiceType -Like Engine | Group-Object InstanceName | Measure-Object Count).Count

### Are you running the latest release?

Yes

### Other details or mentions

_No response_

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe)

### PowerShell Host Version


Name                           Value                                                                                                                                                                                                                             
----                           -----                                                                                                                                                                                                                             
PSVersion                      5.1.19041.1320                                                                                                                                                                                                                    
PSEdition                      Desktop                                                                                                                                                                                                                           
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                                           
BuildVersion                   10.0.19041.1320                                                                                                                                                                                                                   
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                                   
WSManStackVersion              3.0                                                                                                                                                                                                                               
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                               
SerializationVersion           1.1.0.1                                                                                                                                                                                                                           




### SQL Server Edition and Build number

SQL Server 2019, CU15

### .NET Framework Version

.NET Framework 4.8.4470.0

### PR DESCRIPTION
**Pull request by ACALVETT**
## Type of Change

 - [X] Bug fix (non-breaking change, fixes #8249 )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### Purpose
Improve accuracy & quality of the max memory test function

### Approach
Instance counts only apply to Engine and added warning if other components detected.

Added the engine filter on the basis the test only supports engines ATM.
Added warning if other components detected, specifically SSIS,SSAS,SSRS.

### CODE REVIEW

**Comment by ACALVETT** [comment]
Need some input please.

So the tests showed i introduced a divide by 0 issue because of filtering so i put a small tweak in but more is needed however i need some input first.

- Is adding the engine filter the right way to go? I think so ATM 
- If there is no engine is defaulting to 1 the right thing? I think it is as we throw a warning
- I can adjust the tests, i think they failed because the Mock had no engine in it but i could leave them as is since returning a default of 1 addresses the break anyway

Thanks

**Comment by andreasjordan** [comment]
I think it's ok to default to 1 if there is no instance. This way you can use it on an empty box to calculate how the instance should be configured if you would install one.

**Comment by ACALVETT** [comment]
Thanks @andreasjordan, i think from my side this is ready now.

I added a new test and really struggled with the fact connect-dbainstance was throwing a warning, eventually i found some code to make the connect work but it is a blatent copy paste and i don't fully understand it.

File: functions/Test-DbaMaxMemory.ps1
Comment (potatoqualitee) [inline functions/Test-DbaMaxMemory.ps1:96]: Thanks much! Just need a space.

```suggestion
                        Write-Message -Level Warning -Message "The memory calculation may be inaccurate as the following SQL components have also been detected: $($otherConsumers.ServiceType -join(','))"
```

**Comment by potatoqualitee** [comment]
Thank you both! I will merge once @andreasjordan approves.

**Review by andreasjordan — APPROVED** [review]
The code changes look good, but I don't have the time to fully understand the new test. But as it passes, should be ok, too.

**Comment by andreasjordan** [comment]
@potatoqualitee please change "maybe" to "may be" in test test as well.

File: tests/Test-DbaMaxMemory.Tests.ps1
Comment (potatoqualitee) [inline tests/Test-DbaMaxMemory.Tests.ps1:174]:
```suggestion
                $warnvar | Should -BeLike "*The memory calculation may be inaccurate as the following SQL components have also been detected*"
```

**Comment by potatoqualitee** [comment]
done, thank you! will merge and publish once tests run

**Pull request merged**
